### PR TITLE
Holo predev

### PR DESCRIPTION
**Pull request by Ihabilyas**
Multiple Changes, Mainly

1) k_inferred: exposed parameter to infer multiple values up to k, unless prob =0
2) MAP is calculated if k>1 as the value with max prob
3) no more single values and multivalued in accuracy
4) top-k precision and recall measure is the ground truth is contained in our top-k predictions
5) MAP precision and recall as before 
6) Accuracy is simplified now that we don't have single/multi-values
7) Minor fixes to flags of exposed parameters

### CODE REVIEW

File: holoclean/dataset.py [hunk 1]
```diff
@@ -73,7 +73,14 @@ def __init__(self):
...
             'Dimensions_clean': [],
             'Dimensions_dk': [],
-            'Inferred_values': [],
+            'Inferred_values': StructType([
+                    StructField("probability", DoubleType(), False),
+                    StructField("vid", IntegerType(), False),
+                    StructField("attr_name", StringType(), False),
+                    StructField("attr_val", StringType(), False),
+                    StructField("tid", IntegerType(), False),
+                    StructField("domain_id", IntegerType(), False)
```
Comment (thodrek) [inline holoclean/dataset.py:82]: Are these domain_ids universal and is the map stored in dataset? Curiosity.

File: holoclean/holoclean.py [hunk 2]
```diff
@@ -88,25 +88,25 @@
       'default': 0.001,
       'type': float,
       'help': 'The learning rate holoclean will use during training'}),
-    (('-p', '--pruning-threshold1'),
+    (('-pt1', '--pruning-threshold1'),
```
Comment (thodrek) [inline holoclean/holoclean.py:91]: do we want to consider changing to something with reacher meaning. Maybe learning_pruning or something? Not sure.

File: holoclean/holoclean.py [hunk 3]
```diff
@@ -88,25 +88,25 @@
...
      {'metavar': 'PRUNING_THRESHOLD1',
       'dest': 'pruning_threshold1',
       'default': 0.0,
       'type': float,
       'help': 'Threshold1 used for domain pruning step'}),
-    (('-p', '--pruning-threshold2'),
+    (('-pt2', '--pruning-threshold2'),
      {'metavar': 'PRUNING_THRESHOLD2',
       'dest': 'pruning_threshold2',
       'default': 0.3,
```
Comment (thodrek) [inline holoclean/holoclean.py]: I would lower the default value to 0.1

File: holoclean/holoclean.py [hunk 4]
```diff
@@ -136,6 +136,12 @@
       'default': 1,
       'type': int,
       'help': 'The batch size during training'}),
+    (('-ki', '--k-inferred'),
```
Comment (thodrek) [inline holoclean/holoclean.py:139]: if we are not using -k before I'd set the flag to -k

File: holoclean/holoclean.py [hunk 5]
```diff
@@ -809,17 +814,22 @@ def _create_corrected_dataset(self):
...
-            value_predictions = self.inferred_values.collect()
+            # Should not be empty has at least initial values
+            # Persist it to compute MAP as max prediction
+            self.holo_env.dataengine.add_db_table ("Inferred_values",
+                                               self.inferred_values,
+                                               self.dataset)
         else:
-            self.inferred_values = self.simple_predictions
-            value_predictions = self.simple_predictions.collect()
+            self.holo_env.logger.info(" No Inferred_values ")
```
Comment (thodrek) [inline holoclean/holoclean.py:830]: Is this meant to capture an error? If so should we change logging to logger.error or something?

File: holoclean/holoclean.py [hunk 6]
```diff
@@ -809,17 +814,22 @@ def _create_corrected_dataset(self):
...
+            self.holo_env.logger.info("  ")
```
Comment (thodrek) [inline holoclean/holoclean.py:831]: Why do we need this and we don't use \n in the previous message?

File: holoclean/holoclean.py [hunk 7]
```diff
@@ -835,6 +845,39 @@ def _create_corrected_dataset(self):
...
+                                                          "I.attr_name = " \
+                                                          "M.attr_name AND " \
+                                                          "I.probability = " \
+                                                          "M.max_prob"
+
+            self.inferred_map = self.holo_env.dataengine.query(
+                map_inferred_query, 1)
+        else:
+            # MAP is the inferred values when inferred k = 1
+            self.inferred_map = self.inferred_values
```
Comment (thodrek) [inline holoclean/holoclean.py:870]: Want to understand how self.inferred_values is set to the top one from softmax. Or it is the case that it always contains MAP?

File: holoclean/learning/accuracy.py [hunk 8]
```diff
@@ -46,140 +47,101 @@ def accuracy_calculation(self):
...
             init = self.dataengine.query(checkable_original_query, 1)
 
-            if self.session.simple_predictions:
-                inferred_singlevalues = \
-                    self.session.simple_predictions.select("tid", "attr_name",
-                                                           "attr_val")
+            correct_inferred = \
+                inferred.where(inferred.attr_val ==
+                               inferred.g_attr_val).\
+                    drop("attr_val","g_attr_val")
```
Comment (thodrek) [inline holoclean/learning/accuracy.py:67]: why do we need to drop the attribute here?

File: holoclean/learning/accuracy.py [hunk 9]
```diff
@@ -46,140 +47,101 @@ def accuracy_calculation(self):
...
-                else:
-                    singlevalues_recall = 1.0
-
-                singlevalues_precision = float(
-                    inferred_singlevalues_count - incorrect_singlevalues_count
-                ) / inferred_singlevalues_count
-
-                print("The single-value precision that we have is :" + str(
-                    singlevalues_precision))
+            if inferred_count:
```
Comment (thodrek) [inline holoclean/learning/accuracy.py:82]: Where is the else statement? We should print an error message or something.

File: holoclean/learning/accuracy.py [hunk 10]
```diff
@@ -46,140 +47,101 @@ def accuracy_calculation(self):
...
-                print ("The precision that we have is :" + str(precision))
 
-                print ("The recall that we have is :"
-                       + str(recall) + " out of " + str(original_errors_count))
+                print ("The top-" + str(self.session.holo_env.k_inferred) +
+                       " recall is : " + str(recall) + " out of " + str(
+                    incorrect_init_count))
+
+            # Report the MAP accuracy if you are predicting more than 1 value
+            if self.session.holo_env.k_inferred > 1:
```
Comment (thodrek) [inline holoclean/learning/accuracy.py:102]: I thought we already had the MAP values in session.inferred_values isn't that the case? Do we need this query?

File: holoclean/dataset.py (hunk 1 above)
Comment (j48zheng) [reply holoclean/dataset.py:82]: the domain_ids here are the same ones used for possible values, they are saved in the possible_values tables

File: holoclean/holoclean.py (hunk 4 above)
Comment (j48zheng) [reply holoclean/holoclean.py:139]: -k is used for another parameter

File: holoclean/learning/accuracy.py (hunk 10 above)
Comment (j48zheng) [reply holoclean/learning/accuracy.py:102]: I don't think groundtruth is saved in memory. We either use this query or save groundtruth into memory

File: holoclean/learning/accuracy.py [hunk 11]
```diff
@@ -27,14 +27,15 @@ def accuracy_calculation(self):
...
                                        self.dataset.table_specific_name(
                                            'Inferred_Values') + " AS I , " + \
                                        self.dataset.table_specific_name(
-                                           'Groundtruth') + " AS C WHERE " \
-                                                            "C.tid=I.tid " \
+                                           'Groundtruth') + " AS G WHERE " \
+                                                            "I.tid=G.tid " \
                                                             "AND " \
-                                                            "C.attr_name= " \
+                                                            "G.attr_name= " \
```
Comment (ah89) [inline holoclean/learning/accuracy.py:38]: Use better naming for tables

File: holoclean/learning/accuracy.py [hunk 12]
```diff
@@ -46,140 +47,101 @@ def accuracy_calculation(self):
...
-                              ) / original_singlevalues_errors_count)
-                else:
-                    singlevalues_recall = 1.0
-
-                singlevalues_precision = float(
-                    inferred_singlevalues_count - incorrect_singlevalues_count
-                ) / inferred_singlevalues_count
-
-                print("The single-value precision that we have is :" + str(
-                    singlevalues_precision))
```
Comment (ah89) [inline holoclean/learning/accuracy.py:154]: Why this part removed?

**Review by ah89 — APPROVED** [review]

**Review by thodrek — APPROVED** [review]

**Review by j48zheng — APPROVED** [review]
